### ISSUE TITLE
Fill match arm assit

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#626]
We have a powerful-enough type inference to enable **fill match arm** assist! It should turn

```rust
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum TypableDef {
    Function(Function),
    Def(DefId),
}

pub(super) fn type_for_def(db: &impl HirDatabase, def: TypableDef) -> Ty {
    match def {
    }
```

into 

```rust
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum TypableDef {
    Function(Function),
    Def(DefId),
}

pub(super) fn type_for_def(db: &impl HirDatabase, def: TypableDef) -> Ty {
    match def {
        TypableDef::Function(_) => (),
        TypableDef::Def(_) => (),
    }
```

the assist should trigger for both empty arms and a single no-op arm (like the one generated by match postfix completion). 

Not that this assist, unlike all others, makes use of type information, so it should leave in `ra_ide_api` and not in `ra_ide_api_light`.

**Comment by kjeremy** [comment]
Related: It would be super cool to also have a completion for:

```rust
pub(super) fn type_for_def(db: &impl HirDatabase, def: TypableDef) -> Ty {
    match def {
        TypableDef::Function(<|>,
```

I just ran into this with:
```rust
match receiver {
    Ty::Adt {<|>
```

**Comment by gfreezy** [comment]
I'm interested in it. Any hint how to do it?

**Comment by matklad** [comment]
There's no infrastructure in place for this, but it should work like this:

* in ra_ide_api, we should have assits module.
* this module should contain both [this code](https://github.com/rust-analyzer/rust-analyzer/blob/abb9bfe44dfe1fb0685ead5e000ed281c55c8968/crates/ra_ide_api/src/imp.rs#L226-L232) for purely syntactic assists, as well as code for type-based assists.
* type-based assists should have aproximately the same shape as assists in ra_ide_api_light(https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_ide_api_light/src/assists.rs), but they should be given access to `db: &RootDatabase` as well.

**Comment by gfreezy** [comment]
Any more info for how to get the type of the match expression?

**Comment by matklad** [comment]
Here's example code that gets the type: https://github.com/rust-analyzer/rust-analyzer/blob/abb9bfe44dfe1fb0685ead5e000ed281c55c8968/crates/ra_ide_api/src/completion/complete_dot.rs#L8-L18

The workflow is to determine a `hir::Function` at cursor, the call `infer` on it which retunrs InferenceResult which contains types of every expression in the function. Then the type of mathc condition should be extracted and checked if it is an `ENum`.
